### DISCUSSION TITLE
Please reconsider the tech stack for bottles-next

### DISCUSSION THREAD

**Discussion by bitsmurf** [bottlesdevs/Bottles#3641] [General]
Reading the phoronix article and subsequent blog articles regarding bottles-next, I'm deeply concerned with the technology choice for bottles-next and would urge you to reconsider the choice of rust / C# mix as there are better ways to go about the development you want to achieve.
Please allow me to explain:

In your evaluation of technologies, a key player was completely left out and deserves a serious evaluation for the project.

Writing robust GUI applications doesn't have to be difficult, let me introduce you to the Lazarus IDE with a visual GUI designer, a robust component framework which can tap into many underlying UI toolkits from GTK, QT, Cocoa, Win32 or even completely custom drawn and driven by the Modern Object Pascal language and Free Pascal Compiler which is easy to read, easy to learn, really fast to compile and best of all native compiled binaries.
New developer onboarding in the Modern Object Pascal programming languages is a matter of days, rather than weeks that is the case with rust.

Lazarus / Free Pascal has the maturity, many readily available components and libraries to speed up development and a vibrant and dedicated community to help with any questions.
Given the design of the Modern Object Pascal programming language, safety features is already baked in and it is "safe-enough" without fighting the developer.
The write-compile-debug loop is much faster than rust, and the visual GUI designer makes it significantly easier to make adjustments on the fly.
If you're already familiar with Go, Python or C#, Modern Object Pascal will seem quite familiar and easy to grasp.

The best of all, you will be able to have a single code base and environment to take care of the client, server and agent, further speeding up development.

There are docker images available for CI/CD purposes, compilers for many different targets and platforms and I would argue, with the type of application bottles is, 3 - 5 developers would be able to have the rewrite in Modern Object Pascal completed within 6 - 8 months.

Despite what the rust community want one to believe, the language is not fit for this type of application for many good reasons.

Thank you for your reconsideration of this matter.

**Comment by axtloss** [comment]
Bottles is a linux app, we dont need support for win32 or cocoa. Having one codebase for multiple toolkits makes it harder to make a pretty ui since every toolkit has its own quirks. It also doesn't support libadwaita, which we heavily use in bottles.
On top of that, relying on one IDE and forcing everyone to use it just isnt going to work, everyone in the development team has their own setup that they probably dont want to switch away from.
While a visual GUI designer can be useful to make a quick ui, they often produce very bad ui definitions, often adding things like absolute positioning which mess up as soon as the user resizes the window.

**Reply by bitsmurf** [reply]
@axtloss I will respectfully disagree with your multiple wrong assessments, here's why:

Lazarus and the LCL (Lazarus Component Library) transparently maps to the gui toolkit you select, be it GTK3, QT or something else. It does not make it harder to make a pretty UI - on the contrary, it makes it easier. 
They've made a pretty solid abstraction layer which makes sure to select the right toolkit wrapper.
So the quirks you're talking about, you'd have to take it up with the maintainers behind GTK or QT.

As for the question for libadwaita, well, it wouldn't take long to write a wrapper if you're tapping directly into the library to do fancy stuff, but honestly, doing away with libadwaita hard dependency wouldn't be bad.

You don't HAVE to use the Lazarus IDE - there are plugins for VS code and other editors as well. You have many choices to write the code. The IDE just make many things in the development cycle easier.
The way the LCL manages UI elements and windows, it isn't a problem with resizing or scaling of windows.

As said in the initial post, I would suggest you go and play around with Lazarus / Free Pascal first before making uninformed comments about the ecosystem.

**Reply by axtloss** [reply]
I have already used Lazarus and Pascal for multiple projects. Things like the Component Library just aren't that relevant for us since we don't plan on supporting multiple environments at once. Yet another abstraction layer that we partially have to maintain to add support for libadwaita adds more maintenance efforts than just using the library directly. Not to mention that we also want to use iced and libcosmic, which aren't supported by pascal.

**Reply by bitsmurf** [reply]
@axtloss I'm happy to hear you're at least somewhat familiar with Lazarus / Free Pascal, however you're still wrong in your assessment.
Even with Linux being the only target for bottles, the speed and ease of development Lazarus / Free Pascal offers still far outweighs the slight abstraction it will add.
As a matter of fact, it would make it dead easy to support both GTK and QT at the same time, or, project fresnel for a fully custom drawn, css styled and GPU driven UI - so while you're already at a rewrite, do the UI right the first time.

Regarding the route of libcosmic / iced is what is the most concerning of it all, besides being written in rust and has a steep learning curve, it locks developers in, It doesn't offer anything significant, it is a horribly designed ui toolkit requiring an unnecessary amount of work to get anything done. They have committed the same mistake with declarative UI in code as other "modern" toolkits which goes against any good UI composition principles. To make matters worse, it does not integrate well into other DEs than Cosmic, and since the authors behind libcosmic made a poor decision of not exposing a C compatible API for libcosmic to at least allow a level of language freedom, it is doomed to fail.
Whoever sold the idea of rust / libcosmic-iced for bottles - which is not even nowhere near production quality - clearly do not have a thorough understanding of developing solid GUI applications that are easy to maintain.
This seems like a rushed decision made from overhype which is unfortunately the norm in rust land, rather than a solid evaluation of tried, tested and mature technology.

To be frank, this current route will incur a significant and unnecessary burden for development and contributions to an otherwise well developed project like bottles.
It would be sad to see a good project being destroyed by bad tech stack choices.


Also, for those who took the time to thumbs down, why don't you come out of the woodwork and participate in the conversation instead?

**Comment by orowith2os** [comment]
I'm going to close and lock this, because there's no point in this discussion, specifically, occurring. More general discussion around the tech stack being considered is welcome, but this is straight-up refusing to listen. It's not appreciated.
